### PR TITLE
Enhacement | Add biosamples to context list type | #1t2zry3

### PR DESCRIPTION
**Pull request by yairfischer**

### CODE REVIEW

**Comment by nivm** [comment]
Task linked: [CU-1t2zry3 BaseSpace - implement scandir basespace_sdk and basespace_fs api method](https://app.clickup.com/t/1t2zry3)

File: fs_basespace/basespace_context.py [hunk 1]
```diff
@@ -153,8 +156,74 @@ def get_raw_entity_direct(cls, api, sample_id):
         return api.getSampleById(sample_id)
 
 
+class FastQFileGroupContext(CategoryContextDirect):
+    NAME = "fastq files"
```
Comment (amit-ezra) [inline fs_basespace/basespace_context.py:160]: are we sure it's always fastq files? i'm not sure it's the correct name

File: fs_basespace/basespace_context.py [hunk 2]
```diff
@@ -153,8 +156,74 @@ def get_raw_entity_direct(cls, api, sample_id):
...
+    ENTITY_CONTEXT = FileContext
+
+    def list_raw(self, api):
+        # list of raw obj of type DatasetsContext
+        return [element for element in self.raw_obj]
```
Comment (amit-ezra) [inline fs_basespace/basespace_context.py:165]: what type is `self.raw_obj`?
btw you can just `list(self.raw_obj)`

Comment (yairfischer) [reply fs_basespace/basespace_context.py:165]: type: V2DatasetCompact (from sdk)

**Review by amit-ezra — DISMISSED** [review]

**Comment by yairfischer** [comment]
+++++ unit tests coverage: (not yet added to github - need to run vcr)

image

File: fs_basespace/basespace_context.py [hunk 3]
```diff
@@ -153,8 +156,72 @@ def get_raw_entity_direct(cls, api, sample_id):
...
+    ENTITY_ID_FORMAT = re.compile("^ds.[0-9a-z]+$")
+    NAME = "datasets"
+    ENTITY_CONTEXT = SequencedFileGroupsContext
+
+    def list_raw(self, api):
+        return list(self.raw_obj)
+
+    @classmethod
+    def get_raw_entity_direct(cls, api, dataset_id):
+        api_v2_server = 'https://api.basespace.illumina.com/v2'
```
Comment (amit-ezra) [inline fs_basespace/basespace_context.py:185]: please take from the server and not hardcoded

Comment (amit-ezra) [reply fs_basespace/basespace_context.py:185]: `basespace_server`

Comment (amit-ezra) [reply fs_basespace/basespace_context.py:185]: same for all below

File: fs_basespace/basespace_context.py [hunk 4]
```diff
@@ -153,8 +156,72 @@ def get_raw_entity_direct(cls, api, sample_id):
...
+        datasets_api = DatasetsApi(access_token=api.apiClient.apiKey, api_server_and_version=api_v2_server)
```
Comment (amit-ezra) [inline fs_basespace/basespace_context.py:186]: I don't like that we create an API instace and redo the authentication fea each request. It's not ideal. we should use it a session. similar to they way it works with the current API

Comment (amit-ezra) [reply fs_basespace/basespace_context.py:186]: there are severeal ways you can do it that i can think of (maybe you have a better solution)

1. merge this API with the base API so you can use all the methods from the same place. 
2. have a second API instace (for V2) and use it, if merging to the same doesn't work. you can change the base class context to select the correct API for you and save it also under the threading local 
3.

File: fs_basespace/basespace_context.py (hunk 3 above)
Comment (amit-ezra) [reply fs_basespace/basespace_context.py:185]: will be fixed in another PR

File: fs_basespace/basespace_context.py (hunk 4 above)
Comment (amit-ezra) [reply fs_basespace/basespace_context.py:186]: will be fixed in another PR

**Review by amit-ezra — APPROVED** [review]

**Pull request merged**